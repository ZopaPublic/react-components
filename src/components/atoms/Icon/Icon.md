### Summary

Use `<Icon />` whenever you want to render an icon in your application.

This library ( _and the design-system behind_ ) have a strict dependency on the **font-awesome** icon library.

This component is just a wrapper around `<FontAwesomeIcon />` from `'@fortawesome/react-fontawesome'` with just the component name and the `icon` prop renamed for convenience:

```jsx static
// in @fortawesome/react-fontawesome
<FontAwesomeIcon icon={faCoffee} {...otherProps} />

// in @zopauk/react-components
<Icon variant={faCoffee} {...otherProps} />
```

- üßê &nbsp;Check the [`react-fontawesome` docs](https://github.com/FortAwesome/react-fontawesome) for all the possible configuration options.
- üëâüèª &nbsp;You'll need to install `@fortawesome/free-solid-svg-icons` in order to import them explicitly
  - [**explicit imports**](https://github.com/FortAwesome/react-fontawesome#explicit-import) are crucial to prevent big bundles

### Example

```tsx
import { Icon, colors } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<>
  <Icon variant={faCoffee} className="mr-6" />
  <Icon variant={faCoffee} color={colors.brand} className="mr-6" />
  <Icon variant={faCoffee} pulse className="mr-6" />
  <Icon variant={faCoffee} spin className="mr-6" />
  <Icon variant={faCoffee} rotation={180} />
</>;
```

You can also wrap an icon within a circle by setting a `bgColor`

```tsx
import { Icon, colors } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<>
  <Icon variant={faCoffee} bgColor={colors.grey} />
</>;
```
