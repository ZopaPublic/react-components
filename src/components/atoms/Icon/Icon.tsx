import React from 'react';
import styled from 'styled-components';
import { FontAwesomeIcon, FontAwesomeIconProps } from '@fortawesome/react-fontawesome';

import { Colors } from '../../../constants/colors';
import { spacing } from '../../../constants/spacing';
import { IconDefinition } from '@fortawesome/fontawesome-common-types';

export interface IconProps extends Omit<FontAwesomeIconProps, 'icon' | 'border'> {
  variant: IconDefinition;
  bgColor?: Colors[keyof Colors];
}

export const RoundBadge = styled.div<{ color: Colors[keyof Colors] }>`
  background-color: ${({ color }) => color};
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  width: fit-content;
  padding: ${spacing[6]};
`;

export default function Icon({ variant, bgColor, className, ...rest }: IconProps) {
  const renderIcon = (className?: string) => <FontAwesomeIcon {...rest} icon={variant} className={className} />;

  return bgColor ? (
    <RoundBadge className={className} color={bgColor}>
      {renderIcon()}
    </RoundBadge>
  ) : (
    renderIcon(className)
  );
}
