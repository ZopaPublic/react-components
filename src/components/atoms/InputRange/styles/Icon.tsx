import styled from 'styled-components';
import ZopaIcon from '../../Icon/Icon';
import grid from '../../../../constants/grid';

export const Icon = styled(ZopaIcon)`
  display: block;
  font-size: 12px;

  @media (min-width: ${grid.breakpoints.m}px) {
    font-size: 20px;
  }
`;
