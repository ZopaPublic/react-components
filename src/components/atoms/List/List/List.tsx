import React, { HTMLAttributes } from 'react';
import { colors } from '../../../../constants';
import styled from 'styled-components';

const ListComponent = styled.ul<ListProps>`
  padding-left: 28px;
  counter-reset: list-counter;

  & > li:not(.has-icon):before {
    padding: 0;
    left: -25px;
    display: inline-block;
    position: absolute;

    ${({ as = 'ul' }) =>
      as === 'ul'
        ? `
        top: 8px;
      content: '';
      width: 8px;
      height: 8px;
      border-radius: 8px;
      background-color: ${colors.greyDark};
    `
        : `
        top: 1px;
      content: counter(list-counter) ". ";
      color: ${colors.greyDark};
      font-size: 15px;
    `}
  }

  & > li.has-icon svg {
    padding: 0;
    left: -28px;
    top: 4px;
    display: inline-block;
    position: absolute;
    width: 16px;
    height: 16px;
  }
`;

export interface ListProps extends HTMLAttributes<HTMLUListElement | HTMLOListElement> {
  children?: React.ReactNode;
  as?: 'ul' | 'ol';
}

const List = ({ ...props }: ListProps) => <ListComponent {...props} />;

export default List;
