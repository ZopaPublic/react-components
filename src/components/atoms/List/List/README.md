### Summary

Use this component to render a HTML unordered or ordered list of elements with or without icons.

### Usage

```tsx static
<List>
  <List.Item icon={<Icon variant={faCoffee} color={colors.brand} />}>
    <Text as="p">Item One</Text>
  </List.Item>
</List>
```

### Examples

- Accepts icons
- support for ul / ol
- Can be nested

```tsx
import { List, Text, colors, Icon } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<>
  <List>
    <List.Item icon={<Icon variant={faCoffee} color={colors.brand} />}>
      <Text as="p" color={colors.greyDark}>
        Item
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item
      </Text>
      <List>
        <List.Item>
          <Text as="p" color={colors.greyDark}>
            Indented Item one
          </Text>
        </List.Item>
        <List.Item>
          <Text as="p" color={colors.greyDark}>
            Indented Item two
          </Text>
        </List.Item>
        <List.Item>
          <Text as="p" color={colors.greyDark}>
            Indented Item Three
          </Text>
        </List.Item>
      </List>
    </List.Item>
  </List>

  <List as="ol">
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item one
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item two
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item Three
      </Text>
    </List.Item>
  </List>
</>;
```
