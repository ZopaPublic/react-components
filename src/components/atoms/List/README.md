### Summary

Use this component to render a HTML unordered or ordered list of elements with or without icons.

### Examples

- List component can be used as a standard ul / ol:

```tsx
import { List, Text, colors, Icon } from '@zopauk/react-components';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

<>
  <List as="ul">
    <List.Item icon={<Icon variant={faCoffee} color={colors.brand} />}>
      <Text as="p" color={colors.greyDark}>
        Item
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item
      </Text>
    </List.Item>
  </List>

  <List as="ol">
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item one
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item two
      </Text>
    </List.Item>
    <List.Item>
      <Text as="p" color={colors.greyDark}>
        Item Three
      </Text>
    </List.Item>
  </List>
</>;
```
