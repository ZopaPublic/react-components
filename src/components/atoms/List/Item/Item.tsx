import React, { HTMLAttributes, ReactNode } from 'react';
import styled from 'styled-components';
import Icon from '../../Icon/Icon';

export interface ItemProps extends HTMLAttributes<HTMLLIElement> {
  children?: React.ReactNode;
  icon?: typeof Icon | ReactNode;
}

const ItemComponent = styled.li`
  position: relative;
  counter-increment: list-counter;
  list-style: none;
  padding: 0;
`;

const Item = ({ icon, className = '', children, ...rest }: ItemProps) => (
  <ItemComponent className={`${icon ? 'has-icon ' : ''}${className}`} {...rest}>
    {icon}
    {children}
  </ItemComponent>
);

export default Item;
