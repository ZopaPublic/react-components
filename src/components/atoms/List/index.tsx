import React from 'react';

import { default as ListComponent, ListProps } from './List/List';
import Item from './Item/Item';

/**
 * @visibleName List
 */
const List = (props: ListProps) => <ListComponent {...props} />;

export default Object.assign(List, {
  Item,
});
