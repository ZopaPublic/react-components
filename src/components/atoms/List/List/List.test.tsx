import React from 'react';
import { render } from '@testing-library/react';
import Icon from '../../Icon/Icon';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';
import List from './List';
import Item from '../Item/Item';

describe('<List />', () => {
  it('renders as ul', () => {
    const { container } = render(
      <List>
        <li>I am a list item</li>
      </List>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as ol', () => {
    const { container } = render(
      <List as="ol">
        <li>I am a list item</li>
      </List>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with icon', () => {
    const { container } = render(
      <List as="ol">
        <Item icon={<Icon variant={faCoffee} />}>I am a list item</Item>
      </List>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders nested lists', () => {
    const { container } = render(
      <List>
        <Item>
          <p>I am a list item</p>
          <List>
            <Item>
              <p>I am a list item</p>
            </Item>
          </List>
        </Item>
      </List>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
