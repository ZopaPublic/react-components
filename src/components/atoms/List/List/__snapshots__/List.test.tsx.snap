// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> renders as ol 1`] = `
.c0 {
  padding-left: 28px;
  counter-reset: list-counter;
}

.c0 > li:not(.has-icon):before {
  padding: 0;
  left: -25px;
  display: inline-block;
  position: absolute;
  top: 1px;
  content: counter(list-counter) ". ";
  color: #4A545E;
  font-size: 15px;
}

.c0 > li.has-icon svg {
  padding: 0;
  left: -28px;
  top: 4px;
  display: inline-block;
  position: absolute;
  width: 16px;
  height: 16px;
}

<ol
  class="c0"
>
  <li>
    I am a list item
  </li>
</ol>
`;

exports[`<List /> renders as ul 1`] = `
.c0 {
  padding-left: 28px;
  counter-reset: list-counter;
}

.c0 > li:not(.has-icon):before {
  padding: 0;
  left: -25px;
  display: inline-block;
  position: absolute;
  top: 8px;
  content: '';
  width: 8px;
  height: 8px;
  border-radius: 8px;
  background-color: #4A545E;
}

.c0 > li.has-icon svg {
  padding: 0;
  left: -28px;
  top: 4px;
  display: inline-block;
  position: absolute;
  width: 16px;
  height: 16px;
}

<ul
  class="c0"
>
  <li>
    I am a list item
  </li>
</ul>
`;

exports[`<List /> renders nested lists 1`] = `
.c0 {
  padding-left: 28px;
  counter-reset: list-counter;
}

.c0 > li:not(.has-icon):before {
  padding: 0;
  left: -25px;
  display: inline-block;
  position: absolute;
  top: 8px;
  content: '';
  width: 8px;
  height: 8px;
  border-radius: 8px;
  background-color: #4A545E;
}

.c0 > li.has-icon svg {
  padding: 0;
  left: -28px;
  top: 4px;
  display: inline-block;
  position: absolute;
  width: 16px;
  height: 16px;
}

.c1 {
  position: relative;
  counter-increment: list-counter;
  list-style: none;
  padding: 0;
}

<ul
  class="c0"
>
  <li
    class="c1"
  >
    <p>
      I am a list item
    </p>
    <ul
      class="c0"
    >
      <li
        class="c1"
      >
        <p>
          I am a list item
        </p>
      </li>
    </ul>
  </li>
</ul>
`;

exports[`<List /> renders with icon 1`] = `
.c0 {
  padding-left: 28px;
  counter-reset: list-counter;
}

.c0 > li:not(.has-icon):before {
  padding: 0;
  left: -25px;
  display: inline-block;
  position: absolute;
  top: 1px;
  content: counter(list-counter) ". ";
  color: #4A545E;
  font-size: 15px;
}

.c0 > li.has-icon svg {
  padding: 0;
  left: -28px;
  top: 4px;
  display: inline-block;
  position: absolute;
  width: 16px;
  height: 16px;
}

.c1 {
  position: relative;
  counter-increment: list-counter;
  list-style: none;
  padding: 0;
}

<ol
  class="c0"
>
  <li
    class="c1 has-icon "
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-coffee fa-w-20 "
      data-icon="coffee"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M192 384h192c53 0 96-43 96-96h32c70.6 0 128-57.4 128-128S582.6 32 512 32H120c-13.3 0-24 10.7-24 24v232c0 53 43 96 96 96zM512 96c35.3 0 64 28.7 64 64s-28.7 64-64 64h-32V96h32zm47.7 384H48.3c-47.6 0-61-64-36-64h583.3c25 0 11.8 64-35.9 64z"
        fill="currentColor"
      />
    </svg>
    I am a list item
  </li>
</ol>
`;
